Optional outputs

hjoliver opened the pull request:
<!-- Complete this Pull Request template. -->

<!-- Significant PRs should address an existing Issue. Choose one: -->

These changes address https://github.com/cylc/cylc-admin/pull/129

Close #4372 
Close #4256
Close #2914 
Close #3749
Close #4397
Close #4402

- [x]  retrigger incomplete tasks without needing `--reflow` and without merging a new flow number
- [x] Cylc 7 back-compat:
      - suicide triggers still work if needed and have no impact if not needed
      - detect "suite.rc" and print a deprecation warning
      - infer `a:succeed ` and `a:fail` are optional if both appear in the graph
      - make all custom outputs optional (because we can't tell if they should always be completed or not)
- [x] Count incomplete tasks and partially satisfied prereqs toward runahead limit
- [x] If nothing else to run, stall if incomplete tasks or partial prereqs are present 

<!-- The following requirements must be satisfied (with "[x]"). -->
<!-- Mark the PR as a Draft if all requirements are not yet satisfied. -->

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.py` and  `conda-environment.yml`. (None)
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [ ] (master branch) I have opened a documentation PR at cylc/cylc-doc/pull/XXXX.

hjoliver commented:
Not too surprisingly, this branch threw me back into broken-test nightmare-land.  Nearly there, though.

hjoliver commented:
Small demo:
```
[scheduling]
   [[graph]]
      R1 = """
          a & b & c? => d
      """
[runtime]
   [[a]]
   [[b]]
       script = false
   [[c]]
       script = false
   [[d]]
```
Should stall with one incomplete task (`b.1`) and partially satisfied prerequisites for `d.1`:
```
...
2021-08-12T19:00:24+12:00 CRITICAL - [c.1] status=running:
        (received)failed/ERR at 2021-08-12T19:00:23+12:00  for job(01)
        flow(B)
2021-08-12T19:00:24+12:00 CRITICAL - [c.1] -job(01)
        failed
2021-08-12T19:00:24+12:00 INFO - [a.1] status=running: (received)succeeded at
        2021-08-12T19:00:23+12:00  for job(01) flow(B)
2021-08-12T19:00:24+12:00 INFO - (a.1) spawned d.1 flow(B)
2021-08-12T19:00:24+12:00 CRITICAL - [b.1] status=running:
        (received)failed/ERR at 2021-08-12T19:00:23+12:00  for job(01)
        flow(B)
2021-08-12T19:00:24+12:00 CRITICAL - [b.1] -job(01)
        failed
2021-08-12T19:00:24+12:00 WARNING - Incomplete tasks:
        * b.1 (failed) incomplete: succeeded 
2021-08-12T19:00:24+12:00 WARNING - Partially satisfied prerequisites:
        d.1 is waiting on:
        * b.1 succeeded
        * c.1 succeeded
2021-08-12T19:00:24+12:00 CRITICAL - Workflow stalled
```

hjoliver commented:
One integration test failing, one functional test flaky. 

This branch is ready for functional testing, working as per the proposal. I still need to do some code tidying and add more tests before code review.

datamel commented:
@hjoliver I've been running the `tests/functional/api-workflow-info/00-get-graph-raw-1.t` test this end. I can get it to fail (seemingly reliably by deactivating and reactivating my environment), and pass (seemingly with a fresh pip install). It is certainly behaving like a flaky test. @oliver-sanders said he would have a look as he may have a fix on an old branch.

oliver-sanders commented:
It was a total long-shot that the fix required to make the test pass on my platform might also solve this issue, sadly not, but put the fix up anyway - https://github.com/cylc/cylc-flow/pull/4361

The test is highly flaky, about 50:50 for me.

hjoliver commented:
> Ok, think I've got it, the scheduler adds some junk to the get_graph_raw output that isn't sorted.

Yep, that explains it. I merged your test fix and rebase this on to it - all good now, thanks :tada:   And thanks @datamel for taking a look too.

hjoliver commented:
Cylc 7 back-compat done.

Example Cylc 8 workflow branching:

```
[scheduling]
   [[graph]]
      R1 = """
         foo? => post
         foo:fail? => recover
         foo | recover => post
         """
[runtime]
    [[foo, post, recover]]
```
The Cylc 7 version of it:
```
[scheduling]
   [[graph]]
      R1 = """
         foo => post
         foo:fail => recover
         foo | recover => post
         foo => !recover
         foo:fail => !foo
      """
[runtime]
    [[foo, post, recover]]
```
Saving the Cylc 7 version as `flow.cylc`:
```
$ cylc validate flow.cylc
TaskDefError: foo:failed and foo:succeeded can't both be required.
```
Saving as `suite.rc`:
```
$ cylc validate suiter.c
WARNING - Back compat mode turned ON for Cylc 7 "suite.rc" config files.
        Do NOT rename to "flow.cylc" without upgrading to Cylc 8 syntax.
Valid for cylc-8.0b3.dev
```
And it runs successfully, without flagging `foo` as incomplete on success or failure.

hjoliver commented:
Damn, it big conflict with just-merged "infer run number" PR, which I should have realized :facepalm:  I spent quite a bit of time messing with `workflow_files.parse_workflow_arg()` and related code, to handle Cylc 7 back-compat.

hjoliver commented:
Backed off and re-did back-compatibility on the updated branch.

hjoliver commented:
(and broke some tests again :grimacing: )

hjoliver commented on cylc/flow/workflow_files.py:
```diff
@@ -1048,22 +1050,7 @@ def get_platforms_from_db(run_dir):
-
-
-def parse_reg(reg: str, src: bool = False) -> Union[str, Tuple[str, Path]]:
+def parse_reg(reg: str, src: bool = False) -> Tuple[str, Path]:
```
@MetRonnie - not sure if you'll try to talk me out of this. The mypy overloading stuff is cool, but it seems overly complicated to me, for all its worth in this case, especially as mypy was giving me hell trying to adapt this for Cylc 7 back-compat on this branch.  It's more straightforward to have `parse_reg()` always return `(name, path)` and have callers just ignore one element or the other as needed.  What do think?

MetRonnie replied:
Perhaps we should just split it up? Something like `parse_reg(reg: str) -> str` and `parse_src(workflow: str) -> Tuple[str, Path]`. Not sure why I didn't do that in the first place, it seems simple enough unless I've missed something

hjoliver replied:
Yeah we could do that, as a follow-up.

hjoliver commented:
This is done :tada:  - I'll undraft it after checking up on a couple of self-TODOs for potential code tidying

kinow commented on cylc/flow/config.py:
```diff
@@ -1952,16 +1919,61 @@ def load_graph(self):
+                )
+
+    def set_required_outputs(self, graph_parser):
+        """Go through task outupts and set optional/required status."""
```
s/outupts/outputs

kinow reviewed (commented):
Finished reading 107 of 109 files on GitHub UI :+1: great stuff, thanks for the comments @hjoliver . Two big files I will complete tomorrow morning with :coffee: and the IDE to assist. Submitted partial review with a few typos/comments.

oliver-sanders reviewed (commented):
Looks good, a few small comments from the first pass of the code.

MetRonnie reviewed (commented):
Part way through

MetRonnie commented on tests/flakyfunctional/cylc-get-config/04-dummy-mode-output/flow.cylc:
```diff
@@ -11,7 +11,7 @@
     final cycle point = 2000
     [[graph]]
-        P1Y = "foo:meet & bar:greet => baz"
+        P1Y = "foo:meet? & bar:greet? => baz"
```
Oh, because they won't get outputted in the test?

hjoliver replied:
Yes, the live-mode run is not expecting a stall with incomplete tasks.

oliver-sanders commented on cylc/flow/graph_parser.py:
```diff
@@ -65,49 +71,90 @@ class GraphParser:
+        TRIG_FAM_SUBMIT_ANY: (TASK_OUTPUT_SUBMITTED, True),
+    }
+
+    # Map family trigger to member triggers. _ALL True, _ANY False
+    fam_trigger_map = {
+        TRIG_FAM_START_ALL: (TASK_OUTPUT_STARTED, True),
+        TRIG_FAM_START_ANY: (TASK_OUTPUT_STARTED, False),
+        TRIG_FAM_SUCCEED_ALL: (TASK_OUTPUT_SUCCEEDED, True),
+        TRIG_FAM_SUCCEED_ANY: (TASK_OUTPUT_SUCCEEDED, False),
+        TRIG_FAM_FAIL_ALL: (TASK_OUTPUT_FAILED, True),
+        TRIG_FAM_FAIL_ANY: (TASK_OUTPUT_FAILED, False),
+        TRIG_FAM_SUBMIT_ALL: (TASK_OUTPUT_SUBMITTED, True),
+        TRIG_FAM_SUBMIT_ANY: (TASK_OUTPUT_SUBMITTED, False),
+        TRIG_FAM_SUBMIT_FAIL_ALL: (TASK_OUTPUT_SUBMIT_FAILED, True),
+        TRIG_FAM_SUBMIT_FAIL_ANY: (TASK_OUTPUT_SUBMITTED, False),
+        TRIG_FAM_FINISH_ALL: (TASK_OUTPUT_FINISHED, True),
+        TRIG_FAM_FINISH_ANY: (TASK_OUTPUT_FINISHED, False),
+    }
```
It's kinda "infers required", a comment would be nice.

hjoliver replied:
Sorry, I'll add a comment! 

As @oliver-sanders notes, this is inferring default optional/required output status of family members according to the type of family trigger. (Reasoning explained in the proposal).

MetRonnie replied:
But what is the difference with the `fam_output_map` dict above?

Is it this?
> If different family triggers imply both required and optional member output defaults for the same family, we default to optional. You can use the notation without a trigger to force this. E.g.:
>```
># Trigger b only if all members of A succeed, but member success is optional:
>A:succeed-all => b  # default to member success required ...
>A:succeed-any  # ... but now member success is optional
>```

hjoliver replied:
Yikes, no wonder that was confusing, sorry - it's a mess I forgot to go back and clean up. I'll fix it and add some family unit tests.

hjoliver replied:
Actually, the maps are just confusingly named and confusingly similar-looking. One maps family trigger type to member triggers (e.g. `FAM:succeed-all => foo` means `f1:succeed => foo` etc.) and the other maps family trigger type to inferred optionality of member succeed/fail outputs (e.g. `FAM:fail-any` implies `f1:succeed` - and `f1:fail` - is optional, etc.).  I'll try to come up with better names, and add comments...

oliver-sanders commented on cylc/flow/task_pool.py:
```diff
@@ -639,10 +646,11 @@ def remove(self, itask, reason=""):
             self.hidden_pool_changed = True
             if not self.hidden_pool[itask.point]:
                 del self.hidden_pool[itask.point]
+            LOG.debug("[%s] -%s", itask, msg)
```
The hidden pool (and pool implementation in general) is not intended to be user facing. The whole `n=<x>` thing is the user-facing interface which is supposed to divorce task pool implementation from the user interface.

oliver-sanders commented on tests/functional/job-submission/04-submit-num/flow.cylc:
```diff
@@ -5,8 +5,8 @@
         R1 = """
-            foo:fail => bar
-            foo & bar => baz
+            foo:fail? => bar
+            foo? & bar => baz
```
It's a pretty nuts example because `bar` artifically triggers `foo`. Normally `:fail` and `:succeed` would never go hand-in-hand like this.

oliver-sanders commented on tests/functional/triggering/00-recovery/flow.cylc:
```diff
@@ -17,10 +17,10 @@
-           model:fail => diagnose => recover # fail trigger
-           model => !diagnose & !recover  # explicit success trigger
-           model:succeed | recover => post  # conditional and explicit success
+           pre:finish => model?  # finish trigger
+           model:fail? => diagnose => recover # fail trigger
+           model? => !diagnose & !recover  # explicit success trigger
+           model:succeed? | recover => post  # conditional and explicit success
```
I think this comes under "explicit is better than implicit" -- PEP20.

Allowing users to optionally omit the optional output could get more confusing still.

hjoliver replied:
Yeah we discussed this. The idea was that requiring consistency wherever the same task output appears is safer. Otherwise you can't be sure that an output is optional or required without mentally parsing the whole graph.

hjoliver commented on cylc/flow/task_proxy.py:
```diff
@@ -131,8 +127,8 @@ class TaskProxy:
-        .failure_handled:
-            task failure is handled (by children)
+        .incomplete:
+            TODO
```
:+1: - something I backed out of, and neglected to remove it from the doc-string.

hjoliver commented on cylc/flow/taskdef.py:
```diff
@@ -142,6 +150,80 @@ def __init__(self, name, rtcfg, run_mode, start_point, initial_point):
+        else:
+            if old_required in [None, required]:
+                self.outputs[output] = (message, required)
+            elif cylc.flow.flags.cylc7_back_compat:  # noqa: SIM106
```
Hence the warning message:
```
WARNING - Back-compat mode turned ON for Cylc 7 'suite.rc' files.
        Do NOT rename to 'flow.cylc' without upgrading to Cylc 8 syntax.
```

I'll make the 2nd sentence more explicit, something like this:
```
        Do NOT rename your `suite.rc` file to 'flow.cylc' without upgrading to Cylc 8 syntax - it
        is the deprecated file name that turns the back-compat mode on.
```

MetRonnie replied:
What if this scares users away from upgrading the syntax? How about "Upgrade to Cylc 8 syntax BEFORE renaming 'suite.rc' to 'flow.cylc'"

hjoliver replied:
Actually I had already changed it to this, almost exactly as you suggested :grin: :+1: 
```
WARNING - Cylc 7 backward compatibility mode turned ON. Please upgrade
        to Cylc 8 syntax BEFORE renaming your 'suite.rc' to 'flow.cylc'.
        The deprecated filename is what triggers backward compatibility.
```

hjoliver commented on cylc/flow/graph_parser.py:
```diff
@@ -65,49 +71,90 @@ class GraphParser:
-    TRIG_SUCCEED = ':succeed'
-    TRIG_FAIL = ':fail'
-    TRIG_FINISH = ':finish'
-    FAM_TRIG_EXT_ALL = '-all'
-    FAM_TRIG_EXT_ANY = '-any'
-    LEN_FAM_TRIG_EXT_ALL = len(FAM_TRIG_EXT_ALL)
-    LEN_FAM_TRIG_EXT_ANY = len(FAM_TRIG_EXT_ANY)
+    SUICIDE = '!'
+    OPTIONAL = '?'
+    TRIGGER = ':'
+    ARROW = '=>'
+    ACTION = '@'
```
This stuff is only internal, and `@action` wasn't actually introduced into the code here by this branch ... but fair enough, I'll tidy it up...

oliver-sanders replied:
(may impact some error messages further down)

hjoliver commented on cylc/flow/scheduler.py:
```diff
@@ -971,7 +970,7 @@ def set_workflow_timer(self):
             return
         self.workflow_timer_timeout = time() + timeout
-        LOG.debug(
+        LOG.info(
```
I guess it was, but I think it's useful to log when the workflow timer starts counting down.  Note #4372

hjoliver commented on cylc/flow/scheduler.py:
```diff
@@ -1543,7 +1542,7 @@ async def update_data_structure(self):
             if self.workflow_timer_active:
                 self.workflow_timer_active = False
-                LOG.debug(
+                LOG.info(
```
(ditto)

hjoliver commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -292,9 +292,9 @@
-            Conf('abort on timeout', VDR.V_BOOLEAN)
+            Conf('abort on timeout', VDR.V_BOOLEAN, True)
             Conf('abort on inactivity', VDR.V_BOOLEAN)
             Conf('abort on stalled', VDR.V_BOOLEAN)
```
Kind of pending resolution of #4372

hjoliver replied:
(Noted on that issue)

hjoliver replied:
(This PR now implements that).

hjoliver commented on cylc/flow/graph_parser.py:
```diff
@@ -161,6 +209,16 @@ class GraphParser:
         ''', re.X)
 
+    REC_RHS_NODE = re.compile(
+        r'''
+        (!)?                      # optional suicide mark
+        (''' + _RE_NODE + r''')   # node name
+        (''' + _RE_TRIG + r''')?  # optional trigger
+        (''' + _RE_OPT + r''')?   # optional optional indicator
+        ''',
```
Yes that does look nicer :+1:  I'll test tomorrow to make sure we get it right ...

hjoliver commented on tests/functional/triggering/00a-recovery/flow.cylc:
```diff
@@ -0,0 +1,43 @@
+            fi
+        """
+        [[[meta]]]
+            description = A task that succeeds at O UTC and fails at 12 UTC
```
Good spotting! (copied and pasted from `00-recovery.t`, however ... so I'll fix that too)

hjoliver commented on tests/functional/triggering/01-or-conditional/flow.cylc:
```diff
@@ -4,10 +4,10 @@
-            c:failed => !c
+            a | b? => c? & d
+            c? | d => e
+            b:failed? => !b
+            c:failed? => !c
```
Yes, but many of these tests pre-date spawn-on-demand, let alone optional outputs, and I've tended to adapt them minimally because it shows that Cylc 7 style use of suicide triggers still works even though it's not needed anymore (which is required for backward compatibility, of course).  We should make a sweep through all of these tests and fully update them, and then add a small number of well thought-out suicide trigger tests.

hjoliver replied:
#4374

MetRonnie commented on cylc/flow/workflow_files.py:
```diff
@@ -1616,20 +1609,19 @@ def check_flow_file(
             return flow_file_path
         if flow_file_path.resolve() == suite_rc_path.resolve():
             # A symlink that points to *existing* suite.rc
-            if logger:
-                logger.warning(SUITERC_DEPR_MSG)
             return flow_file_path
     if suite_rc_path.is_file():
         if not symlink_suiterc:
-            if logger:
-                logger.warning(SUITERC_DEPR_MSG)
```
I'm not sure these warnings should be removed. Currently there is no warning when doing `cylc play` on a `suite.rc` workflow

hjoliver replied:
I think I've got it working now, for validate and play, with a single log call in the central reg parse function.

MetRonnie commented on tests/unit/test_workflow_files.py:
```diff
@@ -307,14 +306,13 @@ def test_parse_reg__ok(
+        assert Path(expected_path) == parse_reg(reg, src)[1]
     else:
-        expected = expected_reg
-    assert parse_reg(reg, src) == expected
```
I have opened a PR against this PR with suggestions for these tests - https://github.com/hjoliver/cylc-flow/pull/18

hjoliver replied:
Merged, thanks

hjoliver commented:
Note on Cylc 7 back-compat:

I can detect the usual `model:fail` vs `model:succeed` path branching, because `:succeed` and `:fail` are opposite outputs.  So, if this is detected (on this branch) I set `model`s succeed and fail outputs to be optional.
Even without doing this these suites would still run OK, because the original suicide triggers will remove the resulting "incomplete" tasks.  But, setting succeed/fail to be optional means users won't be confronted with a critical error every time `model` finishes.

However, this presents a problem:
```
foo:x => bar  # x and y are mutually exclusive 
foo:y => baz
bar | baz => qux
```

This doesn't need any suicide triggers at Cylc 7, and runs fine on current Cylc 8 master where all outputs are effectively optional.

On this branch (without using the new syntax) `foo` will always be flagged as an incomplete task and can cause a stall.

Not sure what we could do about this except perhaps **assume all custom outputs are optional** in back-compat mode - thoughts?

hjoliver commented:
(looks like I've broken a couple of functional tests now)

MetRonnie commented on cylc/flow/graph_parser.py:
```diff
@@ -85,98 +85,110 @@ class GraphParser:
+    #   "FAM:start-all" to "MEMBER:started" and "-all" (all members).
+    fam_to_mem_trigger_map = {
+        QUAL_FAM_START_ALL: (TASK_OUTPUT_STARTED, True),
+        QUAL_FAM_START_ANY: (TASK_OUTPUT_STARTED, False),
```
Thanks for clarifying this, but I'm still not sure what the boolean means here. Is it literally just `is_all`?

hjoliver replied:
Correct. The alternative to use `.endswith('-all')` on the family qualifier, but I figured a static map is slightly more efficient.

MetRonnie commented on cylc/flow/graph_parser.py:
```diff
@@ -85,98 +85,110 @@ class GraphParser:
+        {_RE_SUICIDE}
+        (?:(?:{TaskID.NAME_RE}             # node name
+        (?:{_RE_PARAMS})?|{_RE_PARAMS}))+  # allow task<param> to repeat
+        (?:{_RE_OFFSET})?                  # optional cycle point offset
+        (?:{RE_QUAL})?                     # optional trigger
+        (?:{_RE_OPT})?                     # optional optional indicator
```
Actually, I don't think it's necessary to state whether the regex group is optional, as that should be reasonably clear from the `?` at the end
```suggestion
        (?:{_RE_OFFSET})?                  # cycle point offset
        (?:{RE_QUAL})?                     # qualifier
        (?:{_RE_OPT})?                     # optional output indicator
```

hjoliver replied:
Good point. I'll make those changes (forgot to change the comments yesterday)

MetRonnie commented on cylc/flow/graph_parser.py:
```diff
@@ -85,98 +85,110 @@ class GraphParser:
+    REC_NODES = re.compile(
+        rf"""
+        ({_RE_NODE_OR_XTRIG})  # node name
+        ({_RE_OFFSET})?        # optional cycle point offset
+        ({RE_QUAL})?           # optional trigger
+        ({_RE_OPT})?           # optional optional indicator
```
```suggestion
        ({_RE_OFFSET})?        # cycle point offset
        ({RE_QUAL})?           # qualifier
        ({_RE_OPT})?           # optional output indicator
```

MetRonnie commented on cylc/flow/graph_parser.py:
```diff
@@ -196,26 +210,30 @@ class GraphParser:
-        (''' + _RE_OPT + r''')?   # optional optional indicator
-        ''',
+        rf"""
+        (!)?           # optional suicide mark
+        ({_RE_NODE})   # node name
+        ({RE_QUAL})?  # optional trigger
+        ({_RE_OPT})?   # optional optional indicator
```
```suggestion
        (!)?           # suicide mark
        ({_RE_NODE})   # node name
        ({RE_QUAL})?   # qualifier
        ({_RE_OPT})?   # optional output indicator
```

MetRonnie commented on cylc/flow/graph_parser.py:
```diff
@@ -458,68 +572,150 @@ def _families_all_to_all(self, expr, rights, info, family_trig_map):
+            # Already in map; check consistency with previous value,
+            if already != optional:
+                if not family:
+                    raise GraphParseError(
+                        f"{name}:{output} can't be both optional and required"
+                    )
```
Codecov is complaining about these - are there not unit tests for these exceptions? (The coverage has been known to misbehave but it should have been fixed by #4348

hjoliver replied:
Oops, more unit tests to come...

MetRonnie commented on cylc/flow/workflow_files.py:
```diff
@@ -320,10 +320,9 @@ class RemoteCleanQueueTuple(NamedTuple):
-    f"Do NOT rename to '{WorkflowFiles.FLOW_FILE}' without "
-    "upgrading to Cylc 8 syntax."
+    f"Deprecated Cylc 7 '{WorkflowFiles.SUITE_RC}' filename detected."
+    f" Upgrade\nto Cylc 8 syntax BEFORE renaming the file "
```
```suggestion
    f" Upgrade to Cylc 8 syntax BEFORE renaming the file "
```

hjoliver replied:
Actually, I can't get nice line breaking in the terminal without newlines in the string. And now I've added more to the message ("Backward compatibility is limited - see Cylc 8 documentation"). So I've gone with newlines between the three sentences.

MetRonnie commented on cylc/flow/workflow_files.py:
```diff
@@ -1075,24 +1074,27 @@ def parse_reg(reg: str, src: bool = False) -> Tuple[str, Path]:
                 f"Workflow name must refer to a directory, not a file: {reg}"
             )
         abs_path, reg = infer_latest_run(abs_path)
+        abs_file_path = check_flow_file(abs_path, logger=None)
```
Hmm, I think `check_flow_file()` will raise if no file is found. But `parse_reg()` shouldn't care whether the file is there or not if `src=False`, as used by "online" commands.

hjoliver replied:
Ah, good point. I'll rethink that.

hjoliver replied:
(I've rediscovered why I was getting double deprecation messages if `check_flow_file` does it - installation results in two checks, one for the installed file, one for the source file).

hjoliver replied:
Found a solution, see what you think.

MetRonnie commented on tests/unit/test_workflow_files.py:
```diff
@@ -321,8 +320,8 @@ def test_parse_reg__ok(
-            None,
-            None,
+            '{cylc_run_dir}/non_exist',
+            "no flow.cylc or suite.rc in ",
```
This was deliberate - there should not be an error if no flow file exists because "online" commands (`src=False`) don't need it
```suggestion
            None,
```

hjoliver replied:
Right, understood now in light of your previous comment!

hjoliver replied:
Restored.

hjoliver commented:
> Not sure what we could do about this except perhaps assume all custom outputs are optional in back-compat mode - thoughts?

I'll go with this for now.

hjoliver commented:
I've addressed all feedback, and the tests all passed, but ... I'm still cleaning up Cylc 7 back-compat.

MetRonnie reviewed (commented):
Slight issue with the `suite.rc` warning now is that it comes before the logo in `cylc play` log, so might be missed
```console
$ cylc play cylc7-simple/ -n --pause
WARNING - Cylc 7 backward compatibility ON for deprecated 'suite.rc' filename.
	Upgrade to Cylc 8 syntax BEFORE renaming the file to 'flow.cylc'.
	Backward compatibility has some limits - see Cylc 8 documentation.

                Cylc Workflow Engine 8.0b3.dev
                Copyright (C) 2008-2021 NIWA
                & British Crown (Met Office) & Contributors

WARNING - deprecated items were automatically upgraded in "workflow definition"
...
```

hjoliver commented on cylc/flow/graph_parser.py:
```diff
@@ -81,38 +90,39 @@ class GraphParser:
+    QUAL_FAM_SUBMIT_ALL: str = "submit-all"
+    QUAL_FAM_SUBMIT_ANY: str = "submit-any"
+    QUAL_FAM_SUBMIT_FAIL_ALL: str = "submit-fail-all"
+    QUAL_FAM_SUBMIT_FAIL_ANY: str = "submit-fail-any"
```
(Yeah I remembered that and stopped doing it just after these lines)

hjoliver commented on cylc/flow/graph_parser.py:
```diff
@@ -616,9 +647,24 @@ def _families_all_to_all(self, expr, rights, info, family_trig_map):
-        """Record parsed triggers and outputs."""
+        self,
+        name: str,
+        output: str,
```
Oops, yes, I moved outputs to a separate method.

hjoliver commented on cylc/flow/graph_parser.py:
```diff
@@ -665,39 +723,48 @@ def _set_output_opt(
+                    reason = self.__class__.FAM_TWEAK
+                LOG.warning(f"{err}\n...{reason}: making it optional.")
+                self.task_output_opt[(name, output)] = (True, is_fam)
+            else:
```
More tests added.

hjoliver commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -300,17 +348,33 @@
             Conf('handlers', VDR.V_STRING_LIST)
             Conf('handler events', VDR.V_STRING_LIST)
             Conf('mail events', VDR.V_STRING_LIST)
-            Conf('startup handler', VDR.V_STRING_LIST)
-            Conf('timeout handler', VDR.V_STRING_LIST)
-            Conf('inactivity handler', VDR.V_STRING_LIST)
-            Conf('shutdown handler', VDR.V_STRING_LIST)
-            Conf('aborted handler', VDR.V_STRING_LIST)
-            Conf('stalled handler', VDR.V_STRING_LIST)
-            Conf('timeout', VDR.V_INTERVAL)
-            Conf('inactivity', VDR.V_INTERVAL)
-            Conf('abort on timeout', VDR.V_BOOLEAN)
-            Conf('abort on inactivity', VDR.V_BOOLEAN)
-            Conf('abort on stalled', VDR.V_BOOLEAN)
```
Note some unplanned config name changes here. I had to distinguish between workflow and stall timeout on this branch, as discussed, so I went slightly further and used consistent naming and grammar for all the timeout and abort items. E.g. `abort on inactivity` really means abort if the inactivity timer times out, hence `abort on inactivity timeout` c.f. `abort on stall timeout` etc.

hjoliver replied:
I'll add them here.

hjoliver commented:
(`remote_background_indep_poll` tests all pass locally)

oliver-sanders reviewed (commented):
One small issue, I think we should flag `finish?` as invalid.

Currently the following validates with no warning, I think the `?` is just ignored?

```ini
[scheduler]                                
    allow implicit tasks = True            
                                           
[scheduling]                               
    [[graph]]                              
        R1 = """                           
            a:finish? => b
        """ 
```

hjoliver commented:
> One small issue, I think we should flag finish? as invalid.

Oops yes, meant to do that. Now done.

oliver-sanders commented:
Going through the family trigger stuff, I've found some examples where cases which would cause a stall at Cylc 7 do not cause a stall at Cylc 8.

Here's my test workflow:

```
[cylc]                                                                                    
    [[parameters]]
        x = 1..12
    [[parameter templates]]
        x = %(x)02d
 
[scheduling]
    cycling mode = integer
    initial cycle point = 1
    [[dependencies]]
        [[[R1]]]
            graph = """
                # run<xx> tasks should all run
                # stall<xx> tasks should all be blocked
 
                GOOD:succeed-all => run01
                GOOD:succeed-any => run02
                GOOD:fail-all => stall01
                GOOD:fail-any => stall02
                GOOD:finish-all => run03
                GOOD:finish-any => run04
 
                BAD:succeed-all => stall03
                BAD:succeed-any => stall04
                BAD:fail-all => run05
                BAD:fail-any => run06
                BAD:finish-all => run07
                BAD:finish-any => run08
 
                UGLY:succeed-all => stall05
                UGLY:succeed-any => run09
                UGLY:fail-all => stall06
                UGLY:fail-any => run10
                UGLY:finish-all => run11
                UGLY:finish-any => run12
            """

[runtime]
    [[GOOD]]  # all pass
    [[BAD]]   # all fail
    [[UGLY]]  # some pass, some fail
 
    [[_good_1, _good_2]]
        inherit = GOOD
        script = true
 
    [[_good_3]]
        inherit = UGLY
        script = true
 
    [[_bad_1, _bad_2]]
        inherit = BAD
        script = false
 
    [[_bad_3]]
        inherit = UGLY
        script = false
 
    [[run<x>, stall<x>]] 
```

### Cylc 7

The `stall<x>` tasks are all stuck in the waiting state and the workflow stalls.

![cylc7](https://user-images.githubusercontent.com/16705946/131655988-c6a1e812-c935-407d-be0f-a265b28df350.png)

### Cylc 8 (with or without compat mode)

Two of the `stall<x>` tasks are spawned and cause a stall due with satisfied prereqs, the other four do not.

```console
$ cylc cat-log custom-outputs/cylc7/family | grep stall
2021-09-01T11:26:35+01:00 WARNING - CYLC 7 BACK-COMPAT ON - deprecated 'suite.rc' filename detected.
	Upgrade to Cylc 8 syntax BEFORE renaming the file to 'flow.cylc'.
	Back-compat mode has some limits - see Cylc 8 documentation.
	
2021-09-01T11:19:53+01:00 INFO - (_bad_3.1) spawned stall06.1 flow(T)
2021-09-01T11:19:53+01:00 INFO - (_good_3.1) spawned stall05.1 flow(T)
	  * stall06.1 is waiting on ['_good_3.1:failed']
	  * stall05.1 is waiting on ['_bad_3.1:succeeded']
2021-09-01T11:20:10+01:00 CRITICAL - Workflow stalled
```

oliver-sanders commented:
## Blocked Task Example

An example from the original cylc-admin proposals:

```
[scheduling]                                                                              
    initial cycle point = 1
    cycling mode = integer
    [[dependencies]]
        [[[R1]]]
            graph = """
                # workflow should stall
                a:x => b    
            """    
                         
[runtime]                
    [[a]]            
        script = true    
        [[[outputs]]]    
            x = x    
    [[b]]    
```

We would expect the workflow to stall with `a` marked as incomplete.

This example stalls as expected under Cylc 7 and Cylc 8, however, at the moment it does not stall when run under Cylc 8 in Cylc 7 compatibility mode.

MetRonnie reviewed (commented):
Happy with all the changes 🎉 (Not explicitly approved pending response to Oliver's comments)

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        This item was previously called ``aborted handler``.
```
```suggestion
           This item was previously called ``aborted handler``.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        Workflow timeout was previously implemented as a stall timeout.
```
```suggestion
           Workflow timeout was previously implemented as a stall timeout.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        Workflow timeout was previously implemented as a stall timeout.
```
```suggestion
           Workflow timeout was previously implemented as a stall timeout.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        Workflow timeout was previously implemented as a stall timeout.
```
```suggestion
           Workflow timeout was previously implemented as a stall timeout.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        This item was previously called ``stalled handler``.
```
```suggestion
           This item was previously called ``stalled handler``.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        This item was previously called ``abort on stalled``.
```
```suggestion
           This item was previously called ``abort on stalled``.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        This item was previously called ``stalled timeout``.
```
```suggestion
           This item was previously called ``stalled timeout``.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        Workflow timeout was previously implemented as a stall timeout.
```
```suggestion
           Workflow timeout was previously implemented as a stall timeout.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        Workflow timeout was previously implemented as a stall timeout.
```
```suggestion
           Workflow timeout was previously implemented as a stall timeout.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        This item was previously called ``inactivity``.
```
```suggestion
           This item was previously called ``inactivity``.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        This item was previously called ``inactivity handler``.
```
```suggestion
           This item was previously called ``inactivity handler``.
```

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,6 +53,130 @@
+
+        .. versionchanged:: 8.0.0
+
+        This item was previously called ``abort on inactivity``.
```
```suggestion
           This item was previously called ``abort on inactivity``.
```

MetRonnie reviewed (commented):
Actually, just built the docs and noticed the `versionchanged` descriptions were below the admonitions, due to under-indenting of the descriptions

hjoliver commented:
@oliver-sanders - I guess the question is, how compatible does back-compat mode need to be exactly? 

```
...
     a:x => b    
...    
[runtime]                
    [[a]]            
        script = true    
        [[[outputs]]]    
            x = x   
``` 

Cylc 7 stalls on this because `b` is pre-spawned and so gets stuck as waiting

Cylc 8 (this branch) stalls on this because `b` finishes as incomplete 

For c7-back-compat we have to infer that custom outputs are optional, because we can't know if they're "required" or not (mutually exclusive custom outputs clearly should not be required) otherwise the following kind of C7 will always stall (when it shouldn't):
```
a:x => b & !c  # x and y are mutually exclusive
a:y => c & !b
```
(Note the suicide triggers don't take out `a` if it gets retained as incomplete).

hjoliver commented:
In your family example, we are correctly inferring that some succeed/fail outputs are optional (they are mutually exclusive by definition, so they must be). In a real c7 suite there should be suicide triggers corresponding to those cases, to remove some of the tasks that you want to cause a stall.

hjoliver commented:
However, I've tried a new approach - the last commit, in back-compat mode, pre-spawns all children when a task is released to be prepared for job submission.  This makes both examples stall exactly like Cylc 7 (with the additional property that the suicide triggers are needed now for real, which I guess is fine). 

Needs some more thought, but this might do the trick...

hjoliver commented:
> Needs some more thought, but this might do the trick...

It seems a bit heavy handed to spawn downstream tasks unnecessarily just to get an entirely Cylc 7-style stall in back-compat mode, but it seems to work fine in all my tests. 

E.g. for your family example above:

```
2021-09-03T15:57:20+12:00 INFO - [run01.1] status=running: (received)succeeded at
        2021-09-03T15:57:19+12:00  for job(01) flow(H)
2021-09-03T15:57:20+12:00 WARNING - Partially satisfied prerequisites:
          * stall04.1 is waiting on ['_bad_2.1:succeeded', '_bad_1.1:succeeded']
          * stall03.1 is waiting on ['_bad_2.1:succeeded', '_bad_1.1:succeeded']
          * stall05.1 is waiting on ['_bad_3.1:succeeded']
          * stall06.1 is waiting on ['_good_3.1:failed']
          * stall02.1 is waiting on ['_good_2.1:failed', '_good_1.1:failed']
          * stall01.1 is waiting on ['_good_2.1:failed', '_good_1.1:failed']
2021-09-03T15:57:20+12:00 CRITICAL - Workflow stalled
2021-09-03T15:57:20+12:00 INFO - PT1H stall timer starts NOW:
        2021-09-03T15:57:20+12:00
````

and the other one:

```
2021-09-03T15:58:35+12:00 WARNING - Partially satisfied prerequisites:
          * bar.1 is waiting on ['foo.1:x']
2021-09-03T15:58:35+12:00 CRITICAL - Workflow stalled
2021-09-03T15:58:35+12:00 INFO - PT1H stall timer starts NOW:
        2021-09-03T15:58:35+12:00
```        

@oliver-sanders - if you're happy with the rest of the PR I think we should merge it as soon as possible to keep the ball rolling.  We can aways tweak back-compat mode later (and  the last commit is small, if we decide to back out of this pre-spawning thing).

hjoliver commented:
OK, pretty sure I've nailed Cylc 7 back-compat mode with final tweaks.

Previously I've said that all outputs are "effectively required" in Cylc 7 because the pre-spawned waiting tasks expect them all to be completed (and have to be removed by suicide triggers if they aren't completed). So I started off by basing back-compat mode on that. That results in workflows that stall in the same circumstances BUT for different reasons: incomplete parent tasks instead of unsatisfied child tasks. Except (as per Oliver's example above) missing custom outputs create a problem because we can't make them required without generating new stalls that are not handled by Cylc 7 suicide triggers (which take out chidren, not incomplete parents, except when used to remove failed tasks).

So, change of tack. In back-compat mode:
- `:submitted` and `:succeeded` are required
  - so failed tasks get retained as incomplete, just as failed tasks in Cylc 7 stay in the pool
- all other outputs are optional
- but children of all outputs are pre-spawned
  - so we get a Cylc 7 style stall even without incomplete parents, and even with single-prerequisite tasks
  - all Cylc 7 suicide triggers will actually be required again, until the workflow is upgraded, to remove these pre-spawned tasks if they are not needed

Pretty sure this covers all the bases and reproduces Cylc 7 behaviour exactly in Cylc 7 suites. It's somewhat annoying to resort to pre-spawning waiting tasks, but it's only in back-compat mode, and being graph-based spawning of the children of active tasks only it is still a truck-load less waiting tasks than Cylc 7 itself generates.

oliver-sanders commented:
This sounds good to me, it's a performance hit, but it only effects the Cylc 7 compat mode and as you say is still better than Cylc 7 itself so it seems like a pretty good compromise to me.

I will rattle through my examples and make sure they behave as expected.

oliver-sanders commented:
Ran through my examples, all good with the exception of the family trigger example (posted above) which now behaves strangely at Cylc 8.

Unfortunately I don't think we can infer the optional outputs in family triggers in this way as an optional output on a family or task in that family anywhere in the graph effectively has the result of transforming a `succeed-all` trigger into a `finish-all` which is definitely not what the user would expect.

Quick example to explain, in this case the `succeed-any` (often used to mean `:succeed-first`) and `succeed-all` can interact:

```
[scheduling]
    [[graph]]
        R1 = """
            A:succeed-all => b
            A:succeed-any => c
        """

[runtime]
    [[A]]
    [[a1, a2]]
        inherit = A
        script = false
    [[b, c]]
```

Because of the "family trigger exemption":

```
2021-09-06T14:54:27+01:00 WARNING - Output c:succeeded is optional so it can't also be required.
	...BUT by the family trigger exemption: making it optional.
2021-09-06T14:54:27+01:00 WARNING - Output b:succeeded is optional so it can't also be required.
	...BUT by the family trigger exemption: making it optional.
2021-09-06T14:54:27+01:00 WARNING - Output b:succeeded is optional so it can't also be required.
	...BUT by the family trigger exemption: making it optional.
2021-09-06T14:54:27+01:00 WARNING - Output c:succeeded is optional so it can't also be required.
	...BUT by the family trigger exemption: making it optional.
```

Which breaks the `succeed-all` trigger causing it to behave like `finish-all`.

Even if all tasks in the family fail the trigger will still fire and the workflow will shut down with no warning:

```
2021-09-06T14:56:31+01:00 CRITICAL - [a1.1] -job(01) failed
2021-09-06T14:56:31+01:00 CRITICAL - [a2.1] -job(01) failed
2021-09-06T14:56:31+01:00 INFO - Workflow shutting down - AUTOMATIC
2021-09-06T14:56:31+01:00 INFO - DONE
```

Note we can end up in the same situation without the `succeed-any` e.g:

```
[scheduling]
    [[graph]]
        R1 = """
            A:succeed-all => b
            a1:fail? => recover
        """
[runtime]
    [[recover]]
        script = """
            cylc broadcast $CYLC_WORKFLOW_NAME <whatever>
            cylc trigger a1.$CYLC_TASK_CYCLE_POINT
        """
```

I'm not sure what the solution is. Perhaps the easiest option is to ditch the attempted optional inference in family triggers for now and consider a new implementation later. This would mean that family triggers would not be compatible with optional branching.

hjoliver commented:
@oliver-sanders - no worries, best to sort this stuff out now!

> Which breaks the `succeed-all` trigger causing it to behave like `finish-all`.

Presumably you do NOT mean that the way it sounds, in terms triggering behaviour? Output optionality does not affect triggering, it only affects whether a finished task is considered complete or not. In your new family example just above, task `b` does not trigger (and it would if the `succeed-all` trigger behaved like `finish-all`).

So you must mean purely in terms of inferring output optionality?  i.e., from `A:succeed-all` you might expect family member success is required, but the presence of `A:succeed-any` makes success optional just like `A:finish-all` would? (But again, it doesn't affect triggering).  I'm not really seeing a problem here, it's we had in the proposal.

(Note I personally preferred a non-inference approach to family triggers, so I'll use this as an excuse to think about that again, but still, I'm not seeing any surprising behaviour in the current implementation).

hjoliver commented:
As commented above, the current way of inferring member output optionality from family triggers seems to me to work fine as per the cylc-admin proposal. However, I've decided I don't like it much for the following reasons:
- inferring different member output optionality for different family triggers may be overly complicated for users, because there's quite a list of family triggers
- clashes of inferred optionality from different triggers can only be handled by making affected outputs optional. This means we can't, for example, have `A:succeed-any => b` with success of all members  *required*, which is a perfectly reasonable thing to want, especially if thinking of an `-any` trigger as "trigger off the first member" which doesn't necessarily imply it is OK for other members to fail.
- this method somewhat conflates output optionality and triggering, which are really orthogonal concepts.

So...

## sub-proposal for handling families better

...

hjoliver commented:
The last commit implements the above, so you can try it.  A lot of tests will break but I can quickly fix that if we agree on this. I'm convinced this is simpler and better.

hjoliver commented:
Just had a VC with @dpmatthews - he's tentatively happy with this new approach to family triggers and optional, we just have to think through and clarify how it works with `:finish` as `:finish-all/any` triggers.

@oliver-sanders - re-read the new proposal above after I try to clarify the text a bit, once again (doing that now...)

oliver-sanders commented:
This sounds good to me, my examples work more or less as expected. My first family trigger example fails validation as Cylc is now detecting the mutually exclusive trigger conditions.

At the moment `succeed-all` and `finish-all` are compatible, however, `fail-all` and `finish-all` aren't which is a little confusing. Otherwise haven't spotted any issues yet.

hjoliver commented:
> At the moment succeed-all and finish-all are compatible, however, fail-all and finish-all aren't which is a little confusing. Otherwise haven't spotted any issues yet.

Yep, I'll fix up the finish-related cases today.

hjoliver commented:
Nailed it, just have to fix the tests.

Don't panic, but I've provisionally gone for requiring a `?` on all `finish` triggers for consistency reasons. E.g.
```
a:finish? => b
  # equivalent to:
a:succeed? | a:fail? => b
```
and:
```
A:finish-all? => b
  # equivalent to:
(a1:succeed? | a1:failed?) & (a2:succeed? | a2:failed?) => b
```

I like this because it means you can think of the `?` as applying to the outputs associated with (/contained within) all triggers.

However, it's a trivial change to relax this constraint if need be, after we debate it some more.

Everything else is straightforward now, I think :relieved:

MetRonnie commented:
Even if it makes sense behind the scenes, I'm not sure requiring `?` for `finish` will make intuitive sense to users. I'd have thought they would expect `finish` to be required, seeing as tasks have to finish one way or another?

hjoliver commented:
> Even if it makes sense behind the scenes, I'm not sure requiring `?` for `finish` will make intuitive sense to users. I'd have thought they would expect `finish` to be required, seeing as tasks have to finish one way or another?

I understand that argument, but I think users need to understand that `finish` is a trigger but not an output (unlike `fail`, etc., which are both) and I'm concerned that pretending otherwise will confuse in other ways. 

Using `a:finish` anywhere means that `a:succeed` and `a:fail` *have to be* made optional elsewhere in the graph, with the `?` syntax, and using it on `finish` too will make that more obvious (think of it as the `?` applies to the *outputs* that contribute to the trigger).

oliver-sanders commented:
I'm guessing we would have to handle the psudo `finished` output at a lower level in order to resolve that.

hjoliver commented:
> Of course we have to explain that `a:finish? => b` does not mean that "finishing is optional" [to trigger b].

Note that this sort of interpretation doesn't work even for the simple task triggers:
```
a:fail? => b
```
This doesn't mean *"failing is optional" in order to trigger `b`*.  In fact failing is *required* to trigger b, but it's an optional output for `a`. We'll really need to hammer this in the documentation.

Makes me think we should revert back to **optional** and **expected** outputs so that we can use required for triggers.

```
a:succeed => b
```
Here, `a:succeed` is an **expected** output of `a` (in terms of consider `a` to be complete or not once finished), but `a:succeed` is **required** in order to trigger `b`.

MetRonnie commented:
> Makes me think we should revert back to **optional** and **expected** outputs so that we can use required for triggers.

Or perhaps "nominal" instead of "expected"? Then again, the word nominal has several different meanings so perhaps not

hjoliver commented:
@dpmatthews - tests now cover implicit success-required if success/fail is not mentioned in the graph. (And, success not required if submit/submit-fail is optional).

Tests all passing. Can I suggest again we try to get this merged as quickly as possible because it has conflict nightmare potential and we need to get working with the new output syntax.  We don't need to resolve the `:finish?` vs `:finish` discussion first, it's a very small follow-up change if we go the other way.

oliver-sanders commented on cylc/flow/graph_parser.py:
```diff
@@ -335,191 +454,409 @@ def _report_invalid_lines(cls, lines):
                 trigs += [
-                    "%s%s%s" % (name, offset, self.__class__.TRIG_SUCCEED),
-                    "%s%s%s" % (name, offset, self.__class__.TRIG_FAIL)]
+                    "%s%s:%s" % (name, offset, TASK_OUTPUT_SUCCEEDED),
+                    "%s%s:%s" % (name, offset, TASK_OUTPUT_FAILED)]
             else:
-                trigs += ["%s%s%s" % (name, offset, trigger)]
+                trigs += [f"{name}{offset}:{trigger}"]
```
Unrelated: Added a low priority issue with a potential memory optimisation - https://github.com/cylc/cylc-flow/issues/4400

MetRonnie commented on cylc/flow/graph_parser.py:
```diff
@@ -335,191 +454,409 @@ def _report_invalid_lines(cls, lines):
+            # Already set; check consistency with previous.
+            if prev_fixed:
+                # optionality fixed already
+                if fam_member:
+                    pass
+                else:
+                    if optional != prev_optional:
+                        raise GraphParseError(
+                            f"Output {name}:{output} can't be both required"
+                            " and optional")
```
```suggestion
                if not fam_member and optional != prev_optional:
                    raise GraphParseError(
                        f"Output {name}:{output} can't be both required"
                        " and optional")
```

MetRonnie commented on cylc/flow/graph_parser.py:
```diff
@@ -335,191 +454,409 @@ def _report_invalid_lines(cls, lines):
+                    if not optional or not opp_default:
+                        raise GraphParseError(
+                            msg + " (via family trigger defaults)")
+                    elif not optional or not opp_optional:
+                        raise GraphParseError(
+                            msg + " (via family trigger)")
```
Would be nice to cover this with a test

hjoliver replied:
Huh, it should be covered already.
```
$ git grep 'via family trigger' tests
tests/unit/test_graph_parser.py:            ("must both be optional if both are used (via family trigger"
tests/unit/test_graph_parser.py:            ("can't default to both optional and required (via family trigger"
```

MetRonnie replied:
Must have been a coverage glitch, the Codecov warning is gone now

MetRonnie commented on tests/flakyfunctional/cylc-show/00-simple/flow.cylc:
```diff
@@ -11,8 +11,8 @@
             bar => foo
-            foo:start => SHOW
-            SHOW:finish-all => end
+            foo:start => SHOW?
+            SHOW:finish-all? => end
```
Not gonna lie I find this a bit confusing. But we can resolve the question of whether `finish` needs a `?` in a follow-up

hjoliver replied:
I'm somewhat expecting to be voted down on this, but let's argue it out to get the best result!

hjoliver replied:
https://github.com/cylc/cylc-flow/issues/4402

oliver-sanders reviewed (approved):
Hard to fully review such a huge and fundamental change quickly, I'm guaranteed to have missed things. Would like to play more but I understand your eagerness :grinning:.

Approving now, I'll keep actively reviewing as long as this is open.

Leaving you to merge / comment / fix / bump to follow on:

* We badly need some Cylc 7 compat mode tests.
  * Especially important for stall logic.
  * I can contribute some of my examples to this cause.
* `abort on timeout` could be `timeout = PT0S` (same for `abort on stall`).
  * Currently `timeout = PT0S` does not behave as expected (same for `stall timeout`).
* Could consider renaming `timeout` to `activity timeout` to better explain it.
* Do `timeout`s get reset? I couldn't get my workflow to timeout twice.

hjoliver commented:
> Hard to fully review such a huge and fundamental change quickly, I'm guaranteed to have missed things. Would like to play more but I understand your eagerness :grinning:.

Yeah, well I've already blown some hours dealing and re-dealing with conflicts between this branch and "integer flow labels".  However, it's most important to get right it right of course, despite the crushing sense of 8.0 release urgency!  I'll leave it unmerged at least till the meeting tonight.

> 
> Approving now, I'll keep actively reviewing as long as this is open.
> 
> Leaving you to merge / comment / fix / bump to follow on:
> 
>     * We badly need some Cylc 7 compat mode tests.
>       
>       * Especially important for stall logic.
>       * I can contribute some of my examples to this cause.

It's well covered in unit tests, in terms of the effect of back-compat mode on task outputs, but yeah I'll add some functional tests now for stall logic. 

> 
>     * `abort on timeout` could be `timeout = PT0S` (same for `abort on stall`).

Ah, I don't follow you.  `abort on timeout` means (duh) abort on the timeout event, after which you can choose to stay up and trigger and event handler instead of aborting.  And, `timeout = PT0S` would (should?) kill the workflow as soon as it starts up, which seems counter-productive to say the least!

 For the stall case, `abort on stall` is redundant with `abort on stall timeout` plus `stall timeout = PT0S`, but I kept both for consistency and because of heavy use of the boolean flag in functional tests. I'll note the equivalence in the doc-strings though...

>       
>       * Currently `timeout = PT0S` does not behave as expected (same for `stall timeout`).

See above, but I'll check ...

> 
>     * Could consider renaming `timeout` to `activity timeout` to better explain it.

IMO not necessary, as it is now essentially equivalent to `timeout 10s cylc play ...`.  If anything, using `workflow timeout` might be best?

> 
>     * Do `timeout`s get reset? I couldn't get my workflow to timeout twice.

That's arguably beyond the scope of this PR, but I'll take a look ...

hjoliver commented:
OK I've fixed zero-interval timeouts. We weren't accounting for the fact that `PT0Y` is false-y.

![shot](https://user-images.githubusercontent.com/2362137/132780103-ced7334e-210a-4c72-b3d7-e9116d6f03f3.png)

hjoliver commented:
The last commit:
- refactors the three workflow timers (they were a mess)
- changes `timeout` to `workflow timeout`, for clarity
- works with zero-interval timeouts

Still TBD:
- tests for the new timeout code
- check on timer reset behaviour
- add Cylc 7 back-compat func tests

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -137,7 +137,7 @@
         .. versionadded:: 8.0.0
 
-           This is how ``timeout handler`` previously behaved.
+           Previously ``timeout`` functioned as a stall timeout.
```
Not sure this is an improvement

hjoliver replied:
On changing the `timeout` item name to `workflow timeout` these docstrings no longer need to explain that `timeout` now means "workflow timeout" whereas it used to mean "stall timeout". Now plain `timeout` is obsolete, so I'm just noting, using this same sentence in every `stall timeout` item that stall timeout functionality used to be configured (Cylc 7) by the `timeout` item. Could still be worded better I'm sure, but we can tweak that later.  Yikes, convoluted.

MetRonnie commented on cylc/flow/scheduler.py:
```diff
@@ -123,16 +124,55 @@ class SchedulerError(CylcError):
+    def reset(self):
+        """Start the timer now (by setting a concrete timeout value)."""
+        self.timeout = time() + self.interval_float
+        self.log_timer_reset(f"{self.interval} {self.name} timer starts NOW")
```
With the `re.sub()` above, "timer" is already in `self.name`
```suggestion
        self.log_timer_reset(f"{self.interval} {self.name} starts NOW")
```

MetRonnie commented on cylc/flow/scheduler.py:
```diff
@@ -123,16 +124,55 @@ class SchedulerError(CylcError):
+        """Stop the timer."""
+        if self.timeout is None:
+            return
+        LOG.warning(f"{self.name} timer stopped")
```
```suggestion
        LOG.warning(f"{self.name} stopped")
```

hjoliver commented:
(Don't bother reviewing more until I've fixed the tests broken by various config changes discussed in the meeting...)

hjoliver commented:
> (Don't bother reviewing more until I've fixed the tests broken by various config changes discussed in the meeting...)

Possibly fixed, with a two-line change, phew.

hjoliver commented:
>    * Do `timeout`s get reset? I couldn't get my workflow to timeout twice.

Yes (now, at least):
- `stall timeout` turns off on any post-stall task activity (e.g. manual triggerIng) and resets on restart (func test added)
- `workflow timeout` resets on restart
- `inactivity timeout` resets on any activity

hjoliver commented:
**Ready for final review** 

Latest changes (as discussed in the meeting and/or just after it) all working now:
- the new family trigger approach
- abstracted workflow timeouts out, and unit tested them (via doctest)
- banned `?` on `finish` triggers
- more functional tests, including for Cylc 7 back-compat stall behaviour
-  `<blargh> handlers` config items now plural for all `<blargh>`  

Test all passing except:
 - `tests/f/cyclers/39-exclusions_advanced.t` fails on macos only, for reasons unknown??
 -  the `_remote_background_indep_poll` batch usually times out (on other branches too)

MetRonnie commented on cylc/flow/cfgspec/globalcfg.py:
```diff
@@ -53,102 +53,106 @@
+        scheduler stalls: if there are no tasks ready to run and no
+        unsatisfied external triggers, but the presence of incomplete
+        tasks or unsatisified prerequisites indicates that the workflow
+        did not run to completion). The stall timer turns off on any
```
Unmatched closing bracket in this sentence

MetRonnie commented on tests/functional/lib/bash/test_header:
```diff
@@ -60,6 +60,8 @@
 #         (stdin if FILE_CONTROL is "-" or missing).
 #     grep_ok PATTERN FILE [$OPTS]
 #         Run "grep [$OPTS] -q -e PATTERN FILE".
+#     grep_workflow_log_ok PATTERN FILE [$OPTS]
```
```suggestion
#     grep_workflow_log_ok TEST_NAME PATTERN [$OPTS]
```

MetRonnie commented on cylc/flow/scheduler.py:
```diff
@@ -711,6 +672,8 @@ def _load_pool_from_point(self):
 
     def _load_pool_from_db(self):
         """Load task pool from DB, for a restart."""
+        LOG.info("Restart from previous run")
+
```
Is this needed? It already logs the restart number
```
INFO - Run: (re)start=4 log=1
```

hjoliver replied:
Nope, you're right. I missed that yesterday because all the reloaded task action timer stuff appears first.  Removing...

hjoliver commented on cylc/flow/cfgspec/workflow.py:
```diff
@@ -289,42 +289,13 @@
                     Conf(item, VDR.V_INTERVAL, desc=desc)
 
-            Conf('abort if startup handler fails', VDR.V_BOOLEAN)
-            Conf('abort if shutdown handler fails', VDR.V_BOOLEAN)
```
We couldn't think of a good reason to keep these. I thought I recalled they were only used in functional tests.  Looking back, I added the first one in 0aeb30b298cfdc9425a8e90890f978cadbe1e756 (16 Aug 2012 :scream: ) along with documentation that says *"This feature is used by cylc test suites which can use shutdown handlers to diagnose the run"*.

hjoliver commented on tests/functional/deprecations/01-cylc8-basic/flow.cylc:
```diff
@@ -19,15 +19,15 @@
-        abort if stalled handler fails =
         abort if timeout handler fails =
+        abort if stalled handler fails =
+        abort if inactivity handler fails =
```
Good spotting, thanks.

hjoliver replied:
(timeout is there already)

hjoliver replied:
> Also at least 1 of the handler -> handlers?

(those are already there too, except for started and shutdown ... adding them)

hjoliver commented on tests/functional/events/24-workflow-timeout/flow.cylc:
```diff
@@ -0,0 +1,12 @@
+        R1 = foo
+[runtime]
+    [[foo]]
+        script = sleep 10
```
Trying to make sure that the workflow times out before it finishes and shuts down normally. But OK, I'll do it properly by making the task watch the log for the timeout...

hjoliver commented:
OK I'm going to merge this in order to free up dependent PRs.
- two approvals
- addressed all feedback
- tests pass, including lots of new tests (and for Cylc 7 back-compat mode)
- 99% patch coverage
